### ISSUE TITLE
M1/Apple Silicon version of Qutebrowser?

### ISSUE THREAD

**Issue by angelus2014** [qutebrowser/qutebrowser#6478]
Hello,

Will there be a M1/Apple Silicon version of Qutebrowser for macOS?

---
Angelo

**Comment by The-Compiler** [comment]
Unlikely, at least for the moment. So far, its dependencies aren't ready (see e.g. https://bugreports.qt.io/browse/QTBUG-85279), but even when they are (if at all, with PyQt and QtWebEngine), I lack the resources to maintain separate M1 builds for a comparatively small part of qutebrowser's users.

**Comment by The-Compiler** [comment]
To elaborate a bit: Happy to reopen this as a wishlist item once the upstream issues are actually taken care of - but right now it's totally unclear whether e.g. PyQt plans to publish wheels (binary releases) for M1. Let's wait for Qt first and see where that goes then.

However chances still are I won't be able to maintain those builds (among other issues, because I don't own the hardware). Once the dependencies are ready, however, you should be able to do `tox -e build-release` to build a release yourself even if I don't do so. But so far, support for macOS and Windows is on a best-effort basis, and I don't see that changing I'm afraid (unless someone trustworthy steps up to take care of macOS stuff and builds).

**Comment by The-Compiler** [comment]
Looks like Qt [will be ready](https://bugreports.qt.io/browse/QTBUG-85279) with Qt 6.2, and PyInstaller is ready with the latest release too: https://github.com/pyinstaller/pyinstaller/pull/5581

Still no concrete plans for official M1 releases at this stage (unless I can build them on a non-M1 Mac!), but at least making it possible to build such a release manually sounds like a good idea.

Let's revisit this once Qt 6.2 is out and qutebrowser supports it (#5395), i.e. somewhen around October or so.

**Comment by shaunsingh** [comment]
Looks like 6.2 is releasing soon. BTW github actions has m1 mac's available, and I would love to build a binary if you ever need

**Comment by The-Compiler** [comment]
> Looks like 6.2 is releasing soon.

Indeed. Note that support for it will first require major changes to qutebrowser (#5395). I did a [quick and dirty approach](https://github.com/qutebrowser/qutebrowser/issues/5395#issuecomment-906544408) in late August which kind-of-worked, but proper support will still take a bit of time I'm afraid.

> BTW github actions has m1 mac's available

It does not: https://github.com/actions/virtual-environments/issues/2187

**Comment by shaunsingh** [comment]
> It does not: actions/virtual-environments#2187
> macos-11
Ah, I see they have macOS big sur images, but those are still intel, sorry for the confusion.
> #5395
Thanks!

**Comment by tgy** [comment]
Hey! What's the current status on the M1-compatible builds? :-)

**Comment by The-Compiler** [comment]
@tgy See the comments above - not much changed about that.

**Comment by stefanpinter** [comment]
hi! does qutebrowser at least work on M1 via emulation/wrapper?

**Comment by The-Compiler** [comment]
@stefanpinter I don't have the hardware to test, but I'm not aware of any M1-specific issues reported.

**Comment by stefanpinter** [comment]
ah thanks @The-Compiler! so why does this issue exist? any problems/cons with going over emulation?

i love qutebrowser

**Comment by The-Compiler** [comment]
Emulation always comes with an overhead - so, less performance than would be possible with a native version, and thus also more battery usage.

**Comment by deividaspetraitis** [comment]
Update: pre-release for github actions for M1 Silicon should be available next week as per this https://github.com/actions/runner/issues/805#issuecomment-1132966327

**Comment by The-Compiler** [comment]
@deividaspetraitis That's only for self-hosted runners, i.e. still means I'd need to buy/rent a M1 Mac to use it.

**Comment by donaldguy** [comment]
FYI to (other) concerned parties, I did finally get the Qt6 branch to build and run (and seems to work, but haven't tested extensively) on my M1 Air.

(took a rundabout journey to get there: https://github.com/qutebrowser/qutebrowser/issues/7202#issuecomment-1156006110)
 but I think its down to just:

```
brew install python # 3.9.13_1, also brings in sqlite @ 3.38.5
gh repo clone qutebrowser/qutebrowser
cd qutebrowser
git checkout qt6-v2
/opt/homebrew/opt/python@3.9/libexec/bin/python scripts/mkvenv.py --pyqt-version 6.3
./.venv/bin/qutebrowser --temp-basedir
```

you might consider throwing a `brew install qt` at the top as well (not so much for the libs as what you need should be pulled from pypi as PyQt6-Qt6 wheel; but maybe for the deps? so far not noticed it being needed but idk)

**Comment by The-Compiler** [comment]
@donaldguy Ah, neat, I didn't realize that PyQt 6 comes with ARM64 wheels already! That certainly gets us a step closer here. The only missing bit - once we have Qt 6 support merged, that is - would then be how to package it all up (ideally without me needing a M1 Mac, or with GitHub Actions having a M1 runner until then).

I'm a bit confused about "This seems to get me back to a working browser!" over in #7202 though: The current .dmg release should work for you as well, other than not running natively and probably #6771, correct?

**Comment by jonathanabennett** [comment]
Once I finish this summer vacation and I'm back on my school-provided Macbook, I am happy to be a tester/compiler for the M1. I'm at least intermediate in Python (been programming in it for 20 years and have 3 small to medium sized projects I've built for schools I've worked at. And I have access to an M1 Silicon Mac. I just chose to bring my Linux laptop on the summer, mostly because I wanted to try setting up and configuring Qutebrowser.

**Comment by david-alvarez-rosa** [comment]
Hi there! My work laptop is M1. I shouldn't be using it too much for personal stuff haha, but I would love to use Qutebrowser at work... if there is something I can do to help, @The-Compiler, please do let me know! 😄

**Comment by The-Compiler** [comment]
@david-alvarez-rosa as outlined above, you should be able to use it just fine without having an M1 build.

**Comment by david-alvarez-rosa** [comment]
Jesus... I have been using Firefox for months... I thought qutebrowser was not working for M1. @The-Compiler this is great news for me! Thank you ❤️

**Comment by DrTobe** [comment]
> FYI to (other) concerned parties, I did finally get the Qt6 branch to build and run (and seems to work, but haven't tested extensively) on my M1 Air.
> 
> ```
> brew install python # 3.9.13_1, also brings in sqlite @ 3.38.5
> gh repo clone qutebrowser/qutebrowser
> cd qutebrowser
> git checkout qt6-v2
> /opt/homebrew/opt/python@3.9/libexec/bin/python scripts/mkvenv.py --pyqt-version 6.3
> ./.venv/bin/qutebrowser --temp-basedir
> ```

I can confirm that @donaldguy's build steps from https://github.com/qutebrowser/qutebrowser/issues/6478#issuecomment-1156016050 work. As far as I can tell, this is already working surprisingly well.

How can I package this into a MacOS application bundle? I guess I could extract the required steps from the Github Actions instructions but it would be easier if this was already written down somewhere.

I could also share the result then if desired.

**Comment by The-Compiler** [comment]
@DrTobe The [releasing scripts](https://github.com/qutebrowser/qutebrowser/blob/master/scripts/dev/build_release.py) currently only work with PyQt from PyPI, so it's not just a matter of running them, they will need changes first to work with Homebrew's Qt.

**Comment by shaggyshoggoth** [comment]
I managed to build an application bundle.

I looked on PyPI and found what seemed to be Apple silicon builds on it. After following @donaldguy 's instructions above, I poked around and edited `tox.ini` to change
`[testenv:pyinstaller-{64,32}{,-qt6}]`
to
`[testenv:pyinstaller-{64,32,arm64}{,-qt6}]`
and then updated `scripts/dev/build_release.sh`
and changed
`call_tox(f'pyinstaller-64{"-qt6" if qt6 else ""}', '-r', debug=debug)`
to
`call_tox(f'pyinstaller-arm64{"-qt6" if qt6 else ""}', '-r', debug=debug)`

Then, from in the top-level git directory I ran
```
source .venv/bin/activate
pip install asciidoc
pip install tox
scripts/dev/build_release.py
```
(The asciidoc may be optional? I did that install long before I had a working build.)

This generated a `dmg` in `dist`, which checks identifies as running on Apple silicon in Activity Monitor.

Note: this is me messing around without knowing exactly what I'm doing.

**Comment by DrTobe** [comment]
> Then, from in the top-level git directory I ran
> 
> ```
> source .venv/bin/activate
> pip install asciidoc
> pip install tox
> scripts/dev/build_release.py
> ```

I guess the last line needs to be `scripts/dev/build_release.py --qt6`?

**Comment by kel50** [comment]
> ```
> brew install python # 3.9.13_1, also brings in sqlite @ 3.38.5
> gh repo clone qutebrowser/qutebrowser
> cd qutebrowser
> git checkout qt6-v2
> /opt/homebrew/opt/python@3.9/libexec/bin/python scripts/mkvenv.py --pyqt-version 6.3
> ./.venv/bin/qutebrowser --temp-basedir
> ```

I can confirm that @donaldguy 's build steps also work for M2 Max.

**Comment by DrTobe** [comment]
> source .venv/bin/activate
> pip install asciidoc
> pip install tox
> scripts/dev/build_release.py

I have tried this again with the recently released `v3.0.0`. Unfortunately, this does not work anymore to get a working app bundle. I have also tried out some similar commands (e.g. the ones found in the Github Actions workflow) but in all cases, the `qutebrowser.app/Contents/MacOS` directory only contained the `qutebrowser` binary. When trying to run it, it complains about not finding `qutebrowser.app/Contents/MacOS/Python` (which actually is not there, but should be in a correct bundle).

Is this expected? Any ideas how to successfully build a bundle on Apple silicon macOS?

**Comment by toofar** [comment]
@DrTobe @hpfmn put a workaround here: https://github.com/qutebrowser/qutebrowser/issues/7859
Hopefully that gets you going for now, not really sure what's going on there but hopefully someone figures it out before the next release.

**Comment by The-Compiler** [comment]
[Introducing the new, Apple silicon powered M1 macOS larger runner for GitHub Actions - The GitHub Blog](https://github.blog/2023-10-02-introducing-the-new-apple-silicon-powered-m1-macos-larger-runner-for-github-actions/)

Will need to see how the pricing thing works, so far I never paid for GitHub Actions usage. I suppose if we only use it for releases (instead of tests) it will be fine either way.

**Comment by toofar** [comment]
Apparently the `macos-14` runner (which is available for public repos too) runs on an M1 CPU now: https://github.com/actions/runner-images/issues/9254

I guess that means if we want to support older macOS releases (which we probably do since I seem to remember people complaining when we drop them) we'll need to installers for mac, one intel for macos11+ and one m1 for macos14+.

Although I'm not actually sure why we couldn't make a universal build with both architectures supported in the x86 runner in the first place. I seem to remember we discussed it and there was some issue, but I can't find that discussion. Apple [says](https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary) you can do that, with a simplified example.

**Comment by The-Compiler** [comment]
Yep, looks like this is [GitHub Actions: Introducing the new M1 macOS runner available to open source! - The GitHub Blog](https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source/)

**Comment by The-Compiler** [comment]
I took a quick look at this today while debugging things for #8226. A few notes:

- Running `tox -e build-release` on a M1 mac does seem to give us an arm64 binary (at least based on the platform shown in `:version`)
- For now, building an `universal2` release (with both architectures in one release) seems to be problematic for two reasons:
  - While (from what it looks like?) PyInstaller can in theory build an arm64 release on a x86_64 Mac, we cannot (smoke) test it.
  - We cannot build such an universal wheel on an M1 machine (where we could smoke-test both architectures thanks to Rosetta2?) either: GitHub Actions only provides M1 runners with macOS 14 Sonoma, however, we need to [build with the oldest supported macOS version](https://pyinstaller.org/en/stable/usage.html#making-macos-apps-forward-compatible), which for now will be macOS 12 Monterey.
  - In practice, it seems to fail with `PyInstaller.utils.osx.IncompatibleBinaryArchError: /Users/user938406/qutebrowser/.tox/pyinstaller/lib/python3.9/site-packages/yaml/_yaml.cpython-39-darwin.so is incompatible with target arch x86_64 (has arch: arm64)!` because PyYAML [does not have universal wheels](https://pypi.org/project/PyYAML/#files).

By default, PyInstaller builds for the currently running architecture, so I think all that'd be left here (as long as we don't mind providing two separate releases rather than one combined one) is some adjustments of CI and release scripts so that we have two macOS machines building/uploading two separate releases.

In the future (once we drop PyYAML, or it upgrades to an universal wheel, or we find a workaround), we could explore building an universal wheel on macOS 12, but then having a separate macOS 14 M1 job that only does the smoke test (and upload).

**Comment by The-Compiler** [comment]
Well, I decided to go all the way and actually make this work - with that, v3.3.0 will have a separate Apple Silicon release. In the meantime, you could [download the latest nightly build](https://github.com/qutebrowser/qutebrowser/actions/workflows/nightly.yml) too.

Universal binaries seem to be a bit more tricky, see #8229.

**Comment by The-Compiler** [comment]
Already released as part of v3.2.1, because it was easier to backport this rather than figuring out how to make the updated GitHub Actions config in `main` work without the changes.

Hope it'll work fine, I only really did a cursory test on Apple Silicon. Feedback welcome!

**Comment by DrTobe** [comment]
I have been using v3.2.1 for a week now. All fine, no crashes, good work, thanks!
